presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/ChunkRecoveryManager.java

```diff
@@ -0,0 +1,427 @@
+        }
+        catch (IOException e) {
+            log.warn(e, "Quarantine of corrupt file failed");
+            tryDeleteFile(file);
+        }
+    }
+
+    private static void deleteFile(Path file)
+    {
+        try {
+            Files.deleteIfExists(file);
+        }
+        catch (IOException e) {
+            throw new PrestoException(RAPTOR_STORAGE_ERROR, "Failed to delete local file: " + file, e);
+        }
+    }
+
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    private static void tryDeleteFile(File file)
+    {
+        file.delete();
+    }
+
+    private static File temporarySuffix(File file)
+    {
+        return new File(file.getPath() + ".tmp-" + randomUUID());
+    }
+
+    private static boolean isFileCorrupt(File file, long size, long xxhash64)
+    {
+        return (file.length() != size) || (xxhash64(file) != xxhash64);
+    }
+
+    private class MissingChunksQueue
+    {
+        private final LoadingCache<MissingChunk, ListenableFuture<?>> queuedMissingChunks;
+
+        public MissingChunksQueue(PrioritizedFifoExecutor<RecoveryRunnable> executor)
+        {
+            requireNonNull(executor, "executor is null");
+            this.queuedMissingChunks = CacheBuilder.newBuilder().build(new CacheLoader<MissingChunk, ListenableFuture<?>>()
+            {
+                @Override
+                public ListenableFuture<?> load(MissingChunk chunk)
+                {
+                    Runnable runnable = () -> restoreFromChunkStore(chunk.getChunkId(), chunk.getSize(), chunk.getXxhash64());
+                    ListenableFuture<?> future = executor.submit(new RecoveryRunnable(chunk.isActive(), runnable));
+                    future.addListener(() -> queuedMissingChunks.invalidate(chunk), directExecutor());
+                    return future;
+                }
+            });
+        }
+
+        public ListenableFuture<?> submit(MissingChunk chunk)
+        {
+            return queuedMissingChunks.getUnchecked(chunk);
+        }
+    }
+
+    private static class RecoveryRunnable
+            implements Runnable
+    {
+        private final boolean active;
+        private final Runnable runnable;
+
+        public RecoveryRunnable(boolean active, Runnable runnable)
+        {
+            this.active = active;
+            this.runnable = requireNonNull(runnable, "runnable is null");
+        }
+
+        @Override
+        public void run()
+        {
+            runnable.run();
+        }
+
+        public static Comparator<RecoveryRunnable> comparator()
+        {
+            return comparingInt(value -> value.active ? 0 : 1);
+        }
+    }
+
+    private static final class MissingChunk
+    {
+        private final boolean active;
+        private final long chunkId;
+        private final long size;
+        private final long xxhash64;
+
+        public MissingChunk(boolean active, long chunkId, long size, long xxhash64)
```

Comment:
"highker on Feb 4  Member
Maybe add static function fromChunk(boolean active, ChunkFile chunk), where chunId, size, and xxhash64 can be derived from chunk.

  @zhaotao1986
zhaotao1986 on Feb 14  Member
will consider it."

I will keep it for now. ChunkFile itself may need refactor later, overlapping with other data structures.